File: app/src/main/java/com/wosloveslife/fantasy/v2/player/State.java
package com.wosloveslife.fantasy.v2.player;

import android.support.annotation.IntDef;

import com.google.android.exoplayer2.ExoPlaybackException;
import com.google.android.exoplayer2.ExoPlayer;
import com.google.android.exoplayer2.Renderer;
import com.google.android.exoplayer2.SimpleExoPlayer;
import com.wosloveslife.fantasy.adapter.ExoPlayerEventListenerAdapter;

import java.util.ArrayList;
import java.util.List;

import static com.wosloveslife.fantasy.v2.player.State.StateCode.STATE_BUFFERING;
import static com.wosloveslife.fantasy.v2.player.State.StateCode.STATE_ENDED;
import static com.wosloveslife.fantasy.v2.player.State.StateCode.STATE_IDLE;
import static com.wosloveslife.fantasy.v2.player.State.StateCode.STATE_READY;

/**
 * 只负责状态变化的传达和当前状态的获取
 * Created by zhangh on 2017/9/3.
 */

public class State {
    @IntDef({STATE_IDLE, STATE_BUFFERING, STATE_READY, STATE_ENDED})
    public @interface StateCode {
        int STATE_IDLE = 1;
        /**
         * The player not able to immediately play from the current position. The cause is
         * {@link Renderer} specific, but this state typically occurs when more data needs to be
         * loaded to be ready to play, or more data needs to be buffered for playback to resume.
         */
        int STATE_BUFFERING = 2;
        /**
         * The player is able to immediately play from the current position. The player will be playing if
         * {@link ExoPlayer#getPlayWhenReady()} returns true, and paused otherwise.
         */
        int STATE_READY = 3;
        /**
         * The player has finished playing the media.
         */
        int STATE_ENDED = 4;
    }

    private List<ExoPlayer.EventListener> mEventListenerBuffers;
    private SimpleExoPlayer mExoPlayer;

    @StateCode
    private int mState;

    void setPlayBinder(SimpleExoPlayer player) {
        mExoPlayer = player;
        if (mExoPlayer == null) {
            return;
        }
            executeAfterBind();

        mExoPlayer.addListener(new ExoPlayerEventListenerAdapter(){
            @Override
            public void onPlayerError(ExoPlaybackException error) {
                super.onPlayerError(error);
                // TODO: 2017/9/24 通知错误
            }
        });
    }

    private void executeAfterBind() {
        if (mEventListenerBuffers != null) {
            List<ExoPlayer.EventListener> mBuffers = new ArrayList<>();
            mBuffers.addAll(mEventListenerBuffers);
            for (ExoPlayer.EventListener listener : mBuffers) {
                addListener(listener);
            }
            mEventListenerBuffers.removeAll(mBuffers);
        }
    }

    public boolean isPlaying() {
        return mExoPlayer != null && mExoPlayer.getPlayWhenReady();
    }

    public long getDuration() {
        return mExoPlayer != null ? mExoPlayer.getDuration() : 0;
    }

    public long getCurrentPosition() {
        return mExoPlayer != null ? mExoPlayer.getCurrentPosition() : 0;
    }

    public long getBufferedPosition() {
        return mExoPlayer != null ? mExoPlayer.getBufferedPosition() : 0;
    }

    public int getPlaybackState() {
        if (mExoPlayer == null) {
            return StateCode.STATE_IDLE;
        }
        return mExoPlayer.getPlaybackState();
    }

    public void badPlay(PlayerException e) {
        // TODO: 2017/9/24 发送通知
    }

    public void setState(int state) {
        mState = state;
    }

    public void addListener(ExoPlayer.EventListener listener) {
        if (mExoPlayer != null) {
            mExoPlayer.addListener(listener);
        } else {
            if (mEventListenerBuffers == null) {
                mEventListenerBuffers = new ArrayList<>();
            }
            mEventListenerBuffers.add(listener);
        }
    }

    public boolean removeListener(ExoPlayer.EventListener listener) {
        boolean remove = false;
        if (mEventListenerBuffers != null) {
            remove = mEventListenerBuffers.remove(listener);
        }
        if (mExoPlayer != null) {
            mExoPlayer.removeListener(listener);
            remove = true;
        }
        return remove;
    }
}

File: app/src/main/java/com/wosloveslife/fantasy/v2/player/State.kt
package com.wosloveslife.fantasy.v2.player

import com.google.android.exoplayer2.ExoPlaybackException
import com.google.android.exoplayer2.Timeline
import com.google.android.exoplayer2.source.TrackGroupArray
import com.google.android.exoplayer2.trackselection.TrackSelectionArray
import com.wosloveslife.dao.Audio
import com.wosloveslife.fantasy.manager.MusicManager
import com.wosloveslife.player.ExoPlayerEventListenerAdapter
import com.wosloveslife.player.IPlayEngine
import com.wosloveslife.player.PlayerException

/**
 * 只负责状态变化的传达和当前状态的获取
 * Created by zhangh on 2017/9/3.
 */
class State {
    companion object State {
        const val STATE_IDLE = 1
        const val STATE_BUFFERING = 2
        const val STATE_READY = 3
        const val STATE_ENDED = 4
    }

    val mEventListener: ArrayList<PlayEvent> = ArrayList()
    private var mPlayer: IPlayEngine? = null

    fun setPlayBinder(player: IPlayEngine) {
        mPlayer = player
        player.addListener(object : ExoPlayerEventListenerAdapter() {
            override fun onTracksChanged(trackGroups: TrackGroupArray, trackSelections: TrackSelectionArray) {
                super.onTracksChanged(trackGroups, trackSelections)
            }

            override fun onLoadingChanged(isLoading: Boolean) {
                super.onLoadingChanged(isLoading)
            }

            override fun onPlayerStateChanged(playWhenReady: Boolean, playbackState: Int) {
                super.onPlayerStateChanged(playWhenReady, playbackState)

                if (playWhenReady) {
                    play(MusicManager.getInstance().musicConfig.mCurrentMusic!!);
                } else {
                    pause()
                }
            }

            override fun onTimelineChanged(timeline: Timeline, manifest: Any) {
                super.onTimelineChanged(timeline, manifest)
            }

            override fun onPositionDiscontinuity() {
                super.onPositionDiscontinuity()
            }

            override fun onPlayerError(error: ExoPlaybackException) {
                super.onPlayerError(error)
                throw PlayerException(PlayerException.FROM_EXO_PLAYER, error)
            }
        })
    }

    private fun play(audio: Audio) {
        for (event in mEventListener) {
            event.onPlay(audio)
        }
    }

    private fun pause() {
        for (event in mEventListener) {
            event.onPause()
        }
    }

    private fun seekTo(progress: Long) {
        for (event in mEventListener) {
            event.onSeekTo(progress)
        }
    }

    fun isPlaying(): Boolean {
        return mPlayer?.isPlaying() ?: false
    }

    fun getDuration(): Long {
        return mPlayer?.getDuration() ?: 0
    }

    fun getProgress(duration: Long): Long {
        return mPlayer?.getCurrentPosition() ?: 0
    }

    fun getBufferProgress(duration: Long): Long {
        return mPlayer?.getBufferedPosition() ?: 0
    }
}